src/core/lib/resource_quota/memory_quota.cc

```diff
@@ -450,6 +501,82 @@ void BasicMemoryQuota::Return(size_t amount) {
   free_bytes_.fetch_add(amount, std::memory_order_relaxed);
 }
 
+uint64_t BasicMemoryQuota::AddNewAllocator(GrpcMemoryAllocatorImpl* allocator) {
+  uint64_t curr_id = curr_allocator_id_.fetch_add(1, std::memory_order_seq_cst);
+
+  AllocatorBucket::Shard& shard = allocators_[0].SelectShard(allocator);
+
+  shard.shard_mu.Lock();
+  shard.allocator_map.insert({curr_id, allocator});
+  shard.shard_mu.Unlock();
+
+  return curr_id;
+}
+
+void BasicMemoryQuota::RemoveAllocator(uint64_t id,
+                                       GrpcMemoryAllocatorImpl* allocator) {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_resource_quota_trace)) {
+    gpr_log(GPR_INFO, "Removing allocator %" PRIu64 ", %p",
+            allocator->allocator_id_, allocator);
+  }
+
+  AllocatorBucket::Shard& small_shard = allocators_[0].SelectShard(allocator);
+  {
+    absl::MutexLock l(&small_shard.shard_mu);
+    if (small_shard.allocator_map.erase(id) == 1) {
+      return;
+    }
+  }
+
+  AllocatorBucket::Shard& big_shard = allocators_[1].SelectShard(allocator);
+
+  big_shard.shard_mu.Lock();
+  big_shard.allocator_map.erase(id);
+  big_shard.shard_mu.Unlock();
+}
+
+void BasicMemoryQuota::MoveAllocatorBigToSmall(
```

Comment:
Suggest renaming to `MaybeMoveAllocatorBigToSmall` -- since there's a conditionality here that would be useful to call out at call sites